state/application.go

```diff
@@ -1484,61 +1487,13 @@ func (a *Application) SetCharm(cfg SetCharmConfig) (err error) {
 			return errors.New("cannot change a charm's deployment mode")
 		}
 	}
-	// For old style charms written for only one series, we still retain
-	// this check. Newer charms written for multi-series have a URL
-	// with series = "".
-	if cfg.Charm.URL().Series != "" {
-		// Allow series change when switching to charmhub charms.
-		if cfg.Charm.URL().Schema != "ch" && cfg.Charm.URL().Series != a.doc.Series {
-			return errors.Errorf("cannot change an application's series")
-		}
-	} else if !cfg.ForceSeries {
-		supported := false
-		charmSeries, err := corecharm.ComputedSeries(cfg.Charm)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		for _, oneSeries := range charmSeries {
-			if oneSeries == a.doc.Series {
-				supported = true
-				break
-			}
-		}
-		if !supported {
-			supportedSeries := "no series"
-			if len(charmSeries) > 0 {
-				supportedSeries = strings.Join(charmSeries, ", ")
-			}
-			return errors.Errorf("only these series are supported: %v", supportedSeries)
-		}
-	} else {
-		// Even with forceSeries=true, we do not allow a charm to be used which is for
-		// a different OS. This assumes the charm declares it has supported series which
-		// we can check for OS compatibility. Otherwise, we just accept the series supplied.
-		currentOS, err := series.GetOSFromSeries(a.doc.Series)
-		if err != nil {
-			// We don't expect an error here but there's not much we can
-			// do to recover.
-			return err
-		}
-		supportedOS := false
-		supportedSeries, err := corecharm.ComputedSeries(cfg.Charm)
+
+	// If it's a v1 or v2 machine charm (no containers), check series.
+	if charm.MetaFormat(cfg.Charm) == charm.FormatV1 || len(cfg.Charm.Meta().Containers) == 0 {
+		err := setCharmCheckSeries(a.doc.Series, cfg.Charm, cfg.ForceSeries)
```

Comment:
reading this at first, it looked like the function called "setCharmCheckseries" actually did some setting and I had to look closer to see that it was just a check.
Can we call it "checkSeriesForSetCharm" or something instead?

Reply:
Done